units/distro_bleed_terminations.go

```diff
@@ -0,0 +1,181 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	hostBleedJobName = "host-bleedoff"
+
+	// if we need to bleed hosts, we want to catch as many hosts as we can that are between jobs
+	idleTimeBleedCutoff = 5 * time.Second
+)
+
+func init() {
+	registry.AddJobType(hostBleedJobName, func() amboy.Job {
+		return makeHostBleedJob()
+	})
+}
+
+type BleedInfo struct {
+	DistroID     string
+	NewCapTarget int
+}
+
+type hostBleedJob struct {
+	job.Base        `bson:"metadata" json:"metadata" yaml:"metadata"`
+	Terminated      int      `bson:"terminated" json:"terminated" yaml:"terminated"`
+	TerminatedHosts []string `bson:"terminated_hosts" json:"terminated_hosts" yaml:"terminated_hosts"`
+
+	env      evergreen.Environment
+	settings *evergreen.Settings
+	host     *host.Host
+
+	BleedInfo BleedInfo
+}
+
+func makeHostBleedJob() *hostBleedJob {
+	j := &hostBleedJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    hostBleedJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	j.SetPriority(2)
+
+	return j
+}
+
+func NewHostBleedJob(env evergreen.Environment, bleedInfo BleedInfo, id string) amboy.Job {
+	j := makeHostBleedJob()
+	j.BleedInfo = bleedInfo
+	j.env = env
+	j.SetID(fmt.Sprintf("%s.%s", hostBleedJobName, id))
+	return j
+}
+
+func (j *hostBleedJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	if j.settings == nil {
+		j.settings = j.env.Settings()
+	}
+
+	if j.HasErrors() {
+		return
+	}
+
+	// get currently existing hosts, in case some hosts have already been terminated elsewhere
+	existingHostCount, err := host.CountRunningHosts(j.BleedInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting existing hosts by Distro.Id"))
+		return
+	}
+	idleHosts, err := host.IdleHostsWithDistroID(j.BleedInfo.DistroID)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "database error getting idle hosts by Distro.Id"))
+		return
+	}
+	bleedTarget := existingHostCount - j.BleedInfo.NewCapTarget
+
+	for i := 0; i < len(idleHosts); i++ {
+		j.AddError(j.checkAndTerminateHost(ctx, &idleHosts[i], &bleedTarget))
+		if bleedTarget == 0 {
+			break
+		}
+	}
+
+	grip.Info(message.Fields{
+		"id":                   j.ID(),
+		"job_type":             hostBleedJobName,
+		"distro_id":            j.BleedInfo.DistroID,
+		"num_idle_hosts":       len(idleHosts),
+		"num_terminated_hosts": j.Terminated,
+		"terminated_hosts":     j.TerminatedHosts,
+	})
+}
+
+func (j *hostBleedJob) checkAndTerminateHost(ctx context.Context, h *host.Host, bleedTarget *int) error {
```

Comment:
A good chunk of this is copied from the hostTerminationJob; can we separate some helper functions maybe?